src/ray/core_worker/reference_count_test.cc

```diff
@@ -1525,6 +1530,88 @@ TEST(DistributedReferenceCountTest, TestDuplicateBorrower) {
+// transited twice on the same worker, a WaitForRefRemoved RPC is still able
+// to retrieve the right containment metadata about the inner id.
+//
+// This unit test covers scenarios from test_dataset.py::test_callable_classes
+// and test_dataset_pipeline.py::test_pipeline_actors.
+//
+// @ray.remote
+// def owner_task1():
+//     inner_id = ray.put(data, _owner=owner)
+//     return inner_id
+//
+// @ray.remote
+// def owner_task2(x):
+//     ray.put(data, _owner=owner)
+//
+// return_id = owner_task1.remote()
+// inner_id = ray.get(outer_id)[0]
+// return_id2 = owner_task2.remote(inner_id)
+//
+TEST(DistributedReferenceCountTest, TestForeignOwner) {
+  auto caller = std::make_shared<MockWorkerClient>("1");
+  auto owner = std::make_shared<MockWorkerClient>("2");
+  auto foreign_owner = std::make_shared<MockWorkerClient>("3");
+
+  //
+  // Phase 1 -- submit and execute owner_task1()
+  //
+  // Caller submits a task.
+  auto return_id = caller->SubmitTaskWithArg(ObjectID::Nil());
+  // Task returns inner_id as its return value.
+  auto inner_id = ObjectID::FromRandom();
+  owner->PutWithForeignOwner(inner_id, foreign_owner->address_);
+  rpc::WorkerAddress addr(caller->address_);
+  auto refs = owner->FinishExecutingTask(ObjectID::Nil(), return_id, &inner_id, &addr);
+  ASSERT_TRUE(refs.empty());
+  ASSERT_TRUE(owner->rc_.HasReference(inner_id));
+  // Caller receives the owner's message, but inner_id is still in scope
+  // because caller has a reference to return_id.
+  caller->HandleSubmittedTaskFinished(ObjectID::Nil(), {{return_id, {inner_id}}});
+
+  //
+  // Phase 2 -- submit and execute owner_task2(x)
+  //
+  auto return_id2 = caller->SubmitTaskWithArg(return_id);
+  caller->rc_.RemoveLocalReference(return_id, nullptr);
+  ASSERT_TRUE(owner->rc_.HasReference(inner_id));
+  caller->rc_.RemoveLocalReference(return_id2, nullptr);
+  // Owner receives a reference to inner_id. It still has a reference when
+  // the task returns.
+  owner->ExecuteTaskWithArg(return_id, inner_id, owner->address_);
+  auto refs2 = owner->FinishExecutingTask(return_id, return_id);
+  // owner merges ref count into the caller.
+  caller->HandleSubmittedTaskFinished(return_id, {}, owner->address_, refs2);
+  ASSERT_FALSE(owner->rc_.HasReference(return_id));
+  ASSERT_FALSE(caller->rc_.HasReference(return_id));
+  ASSERT_FALSE(owner->rc_.HasReference(return_id2));
+  ASSERT_FALSE(caller->rc_.HasReference(return_id2));
+  ASSERT_TRUE(owner->rc_.HasReference(inner_id));
+
+  //
+  // Phase 3 -- foreign owner gets ref removed information.
+  //
+  // Emulate ref removed callback.
+  ReferenceCounter::ReferenceTableProto refs_proto;
+  owner->rc_.GetAndClearLocalBorrowersInternal(inner_id, /*for_ref_removed=*/true,
+                                               &refs_proto);
+  ASSERT_TRUE(owner->rc_.HasReference(inner_id));
+
+  // This is the key assert checking the info is still available.
+  ASSERT_EQ(refs_proto[0].stored_in_objects().size(), 1);
```

Comment:
This assertion is implementation-dependent. Can we change it to check the higher-level invariant: the foreign owner still has the ref in scope if the caller has the ref in scope (right now it looks like the caller is deleting its ref, so the race condition doesn't cause a bug).